bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
+calc_load(unsigned long load, unsigned long exp, unsigned long active)
+{
+	unsigned long newload;
+
+	active = active > 0 ? active * FIXED_1 : 0;
+	newload = load * exp + active * (FIXED_1 - exp);
+	if (active >= load)
+		newload += FIXED_1 - 1;
+
+	return newload / FIXED_1;
+}
+
+/*
+ * Return 0 means that container p->cg is closed.
+ * Return -1 means that error occurred in refresh.
+ * Positive num equals the total number of pid.
+ */
+static int refresh_load(struct load_node *p, char *path)
+{
+	FILE *f = NULL;
+	char **idbuf;
+	char proc_path[256];
+	int i, ret, run_pid = 0, total_pid = 0, last_pid = 0;
+	char *line = NULL;
+	size_t linelen = 0;
+	int sum, length;
+	DIR *dp;
+	struct dirent *file;
+
+	idbuf = (char **)malloc(sizeof(char *));
+	if (!idbuf) {
+		lxcfs_error("Refresh container:%s failed for malloc!\n", p->cg);
+		return -1;
+	}
+	sum = calc_pid(&idbuf, path, DEPTH_DIR, 0, p->cfd);
+	/*  normal exit and abnormal exit */
+	if (sum == 0 || sum == -1)
+		goto out;
+
+	for (i = 0; i < sum; i++) {
+		/*clean up '\n' */
+		length = strlen(idbuf[i])-1;
+		idbuf[i][length] = '\0';
+		do {
+			ret = snprintf(proc_path, 256, "/proc/%s/task", idbuf[i]);
+		} while (ret < 0);
+
+		dp = opendir(proc_path);
+		if (!dp) {
+			lxcfs_error("%s\n", "Open proc_path failed in refresh_load.");
+			continue;
+		}
+		while ((file = readdir(dp)) != NULL) {
+			if (strncmp(file->d_name, ".", 1) == 0)
+				continue;
+			total_pid++;
+			/* We make the biggest pid become last_pid.*/
+			ret = atof(file->d_name);
+			last_pid = (ret > last_pid) ? ret : last_pid;
+
+			do {
+				ret = snprintf(proc_path, 256, "/proc/%s/task/%s/status", idbuf[i], file->d_name);
+			} while (ret < 0);
+			f = fopen(proc_path, "r");
+			if (f != NULL) {
+				while (getline(&line, &linelen, f) != -1) {
+					/* Find State */
+					if ((line[0] == 'S') && (line[1] == 't'))
+						break;
+				}
+			if ((line[7] == 'R') || (line[7] == 'D'))
+				run_pid++;
+			fclose(f);
+			}
+		}
+		closedir(dp);
+	}
+	/*Calculate the loadavg.*/
+	p->avenrun[0] = calc_load(p->avenrun[0], EXP_1, run_pid);
+	p->avenrun[1] = calc_load(p->avenrun[1], EXP_5, run_pid);
+	p->avenrun[2] = calc_load(p->avenrun[2], EXP_15, run_pid);
+	p->run_pid = run_pid;
+	p->total_pid = total_pid;
+	p->last_pid = last_pid;
+
+	free(line);
+	for (; i > 0; i--)
+		free(idbuf[i-1]);
+out:
+	free(idbuf);
+	return sum;
+}
+/*
+ * Traverse the hash table and update it.
+ */
+void *load_begin(void *arg)
+{
+
+	char *path = NULL;
+	int i, sum, length, ret;
+	struct load_node *f;
+	int first_node;
+	clock_t time1, time2;
+
+	while (1) {
+		time1 = clock();
+		for (i = 0; i < LOAD_SIZE; i++) {
+			pthread_mutex_lock(&load_hash[i].lock);
+			if (load_hash[i].next == NULL) {
+				pthread_mutex_unlock(&load_hash[i].lock);
+				continue;
+			}
+			f = load_hash[i].next;
+			first_node = 1;
+			while (f) {
+				length = strlen(f->cg) + 2;
+				do {
+					/* strlen(f->cg) + '.' or '' + \0 */
+					path = (char *)malloc(length);
```

Comment:
Please don't cast the result of `malloc()`.
Can you explain why you do a looping `malloc()` here? I mean it's fine as this will be killed by the OOM kill but I'd like to know why here and not in other places.

Reply:
At first I want to use loops in all malloc() situations. But I don’t know whether it is safe. 
It's hard to say why I do a looping malloc() here.
I will use loops in all malloc() situations.
Thank you!